### PR TITLE
Short term way to make `AggregateStatistics` still work when min/max is converted to udaf

### LINKED ISSUE #11153: Remove `Min`/`Max` references from `AggregateStatistics`
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/datafusion/issues/11151 (where we are removing special case uses of Min/Max)

### Describe the solution you'd like

Reemove these cases:

https://github.com/apache/datafusion/blob/569be9eb1ba60d3e842f22de5080c732ac2206f4/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L185-L186

https://github.com/apache/datafusion/blob/569be9eb1ba60d3e842f22de5080c732ac2206f4/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L197-L198

https://github.com/apache/datafusion/blob/569be9eb1ba60d3e842f22de5080c732ac2206f4/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L235-L236

https://github.com/apache/datafusion/blob/569be9eb1ba60d3e842f22de5080c732ac2206f4/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L247-L248

Specifically, the idea is to remove the pattern

```rust
       expr.as_any().downcast_ref::<Max>()
```

and 

```rust
       expr.as_any().downcast_ref::<Min>()
```

### Describe alternatives you've considered

I suggest adding a general purpose function to `AggregateExec` and then implementing it for Min and Max (so we can do the same for Min/Max UDAF)

```rust
impl AggregateExpr {

 /// Return the value of the aggregate function, if known, given the number of input rows.
 /// 
 /// Return None if the value can not be determined solely from the input.
 /// 
 /// # Examples
 /// * The `COUNT` aggregate would return `Some(11)` given `num_rows = 11`
 /// * The `MIN` aggregate would return `Some(Null)` given `num_rows = 0
 /// * The `MIN` aggregate would return `None` given num_rows = 11
 fn output_from_rows(&self, num_rows: usize) -> Option<ScalarValue> { None }
...
}

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Rachelint**
## Which issue does this PR close?

Part of  #11153 

## Rationale for this change
Now only hard code like `if agg_expr.as_any().is::<expressions::Min>()` to identify the `min/max` aggregate function in `AggregateStatistics` optimizer. It can't work after `min/max` is converted to `udaf`.

This pr is for solving problem stated above. The best way to solve is adding a function like `output_from_stats` into `AggregateExpr` and moving the logic from the optimizer to `udaf`. But unfortunately it is not so easy to find good way to do that... 

For unblocking the `udaf` conversion about `min/max`, just impl a short term solution here.

## What changes are included in this PR?
Support to identify `min/max` after they are converted to `udaf` in later.

## Are these changes tested?
By exist tests.

## Are there any user-facing changes?
No.

### CODE REVIEW

**Comment by Rachelint** [comment]
@alamb @EduardoVega 
I have made draf impl...
I define a new function `output_from_stats` as a general function to get results directly from stats if possible.
Howerver, when I impl adapt it to `AggregateUDFImpl`, I found I can't get the physical exprs which is needed, see the original code in physical optimizer:
https://github.com/apache/datafusion/blob/569be9eb1ba60d3e842f22de5080c732ac2206f4/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L250-L257

It is mainly due to `AggregateUDFImpl` is defined in `expr` crate, and the needed `PhysicalExpr` is defined in `datafusion-physical-expr-common` crate which will refer `exps` a dependency. And if we import `PhysicalExpr` in `expr`, that will lead to cyclic dependecies...

For solving it, I define a `AggregateArgExprs` trait in `expr` crate, and impl it for the `Vec<Arc<dyn PhysicalExpr>>`. It can work but we need to downcast it to `&[Arc<dyn PhysicalExpr>]` when using it in `output_from_stats` of the `udaf`(see the impl of `count`...). I am not sure if it is a good way to impl like this...

**Comment by alamb** [comment]
Thank you @Rachelint ❤️ 

 > It is mainly due to AggregateUDFImpl is defined in expr crate, and the needed PhysicalExpr is defined in datafusion-physical-expr-common crate which will refer exps a dependency. And if we import PhysicalExpr in expr, that will lead to cyclic dependecies...

This makes sense -- I agree the ` AggregateUDFImpl` API should not be in terms of `PhysicalExpr` (for the dependency reasons you describe)

One way I could think to get around this would be to figure out the column statistics in the physical planner and pass it directly (rather than making the UDF figure them out). Something like

```rust
pub trait AggregateUDFImpl {
...
   /// If the output of this aggregate function can be determined 
   /// only from input statistics (e.g. COUNT(..) with 0 rows, is always 0)
   /// return that value
   ///
   /// # Arguments
   /// stats: Overall statistics for the input (including row count)
   /// arg_stats: Statistics, if known, for each input argument
    fn output_from_stats(
        &self,
        stats: &Statistics,
        arg_stats: &[Option<&ColumnStatistics>]
    ) -> Option<ScalarValue> {
        None
    }
...
}
```

ANother option might be to use the narrower API described on https://github.com/apache/datafusion/issues/11153

```rust
impl AggregateExpr {

 /// Return the value of the aggregate function, if known, given the number of input rows.
 /// 
 /// Return None if the value can not be determined solely from the input.
 /// 
 /// # Examples
 /// * The `COUNT` aggregate would return `Some(11)` given `num_rows = 11`
 /// * The `MIN` aggregate would return `Some(Null)` given `num_rows = 0
 /// * The `MIN` aggregate would return `None` given num_rows = 11
 fn output_from_rows(&self, num_rows: usize) -> Option<ScalarValue> { None }
...
}
```

Though I think your idea is better (pass in statistics in generla)

**Comment by edmondop** [comment]
The draft implementation seems reasonable and all checks are passing 🎉

**Comment by Rachelint** [comment]
> Thank you @Rachelint ❤️
> 
> > It is mainly due to AggregateUDFImpl is defined in expr crate, and the needed PhysicalExpr is defined in datafusion-physical-expr-common crate which will refer exps a dependency. And if we import PhysicalExpr in expr, that will lead to cyclic dependecies...
> 
> This makes sense -- I agree the ` AggregateUDFImpl` API should not be in terms of `PhysicalExpr` (for the dependency reasons you describe)
> 
> One way I could think to get around this would be to figure out the column statistics in the physical planner and pass it directly (rather than making the UDF figure them out). Something like
> 
> ```rust
> pub trait AggregateUDFImpl {
> ...
>    /// If the output of this aggregate function can be determined 
>    /// only from input statistics (e.g. COUNT(..) with 0 rows, is always 0)
>    /// return that value
>    ///
>    /// # Arguments
>    /// stats: Overall statistics for the input (including row count)
>    /// arg_stats: Statistics, if known, for each input argument
>     fn output_from_stats(
>         &self,
>         stats: &Statistics,
>         arg_stats: &[Option<&ColumnStatistics>]
>     ) -> Option<ScalarValue> {
>         None
>     }
> ...
> }
> ```
> 
> ANother option might be to use the narrower API described on #11153
> 
> ```rust
> impl AggregateExpr {
> 
>  /// Return the value of the aggregate function, if known, given the number of input rows.
>  /// 
>  /// Return None if the value can not be determined solely from the input.
>  /// 
>  /// # Examples
>  /// * The `COUNT` aggregate would return `Some(11)` given `num_rows = 11`
>  /// * The `MIN` aggregate would return `Some(Null)` given `num_rows = 0
>  /// * The `MIN` aggregate would return `None` given num_rows = 11
>  fn output_from_rows(&self, num_rows: usize) -> Option<ScalarValue> { None }
> ...
> }
> ```
> 
> Though I think your idea is better (pass in statistics in generla)



> Thank you @Rachelint ❤️
> 
> > It is mainly due to AggregateUDFImpl is defined in expr crate, and the needed PhysicalExpr is defined in datafusion-physical-expr-common crate which will refer exps a dependency. And if we import PhysicalExpr in expr, that will lead to cyclic dependecies...
> 
> This makes sense -- I agree the ` AggregateUDFImpl` API should not be in terms of `PhysicalExpr` (for the dependency reasons you describe)
> 
> One way I could think to get around this would be to figure out the column statistics in the physical planner and pass it directly (rather than making the UDF figure them out). Something like
> 
> ```rust
> pub trait AggregateUDFImpl {
> ...
>    /// If the output of this aggregate function can be determined 
>    /// only from input statistics (e.g. COUNT(..) with 0 rows, is always 0)
>    /// return that value
>    ///
>    /// # Arguments
>    /// stats: Overall statistics for the input (including row count)
>    /// arg_stats: Statistics, if known, for each input argument
>     fn output_from_stats(
>         &self,
>         stats: &Statistics,
>         arg_stats: &[Option<&ColumnStatistics>]
>     ) -> Option<ScalarValue> {
>         None
>     }
> ...
> }
> ```
> 
> ANother option might be to use the narrower API described on #11153
> 
> ```rust
> impl AggregateExpr {
> 
>  /// Return the value of the aggregate function, if known, given the number of input rows.
>  /// 
>  /// Return None if the value can not be determined solely from the input.
>  /// 
>  /// # Examples
>  /// * The `COUNT` aggregate would return `Some(11)` given `num_rows = 11`
>  /// * The `MIN` aggregate would return `Some(Null)` given `num_rows = 0
>  /// * The `MIN` aggregate would return `None` given num_rows = 11
>  fn output_from_rows(&self, num_rows: usize) -> Option<ScalarValue> { None }
> ...
> }
> ```
> 
> Though I think your idea is better (pass in statistics in generla)

One thing I worry about the narrow api is that, it seems can't be used to support the origin optimization of min/max?
https://github.com/apache/datafusion/blob/7df000a333a7d4018e1446ef900f652288b1f104/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L195-L221

Maybe I misunderstand about it?

**Comment by alamb** [comment]
> @alamb One thing I worry about the narrow api is that, it seems can't be used to support the original optimization of min/max?
> 
> Maybe I misunderstand about it?

No, sorry, you are correct, I was mistaken

**Comment by Rachelint** [comment]
> > @alamb One thing I worry about the narrow api is that, it seems can't be used to support the original optimization of min/max?
> > Maybe I misunderstand about it?
> 
> No, sorry, you are correct, I was mistaken

Ok... Been a bit busy the past couple of days, continue to read the related codes and think a relative good way to solve this today...
One way I could think to solve it temporarily for not blocking the progress of #10943 
is that we just treat it a special case like `count`:
https://github.com/apache/datafusion/blob/1e0c06e14ae821ac6aa344f8acb638431a898ae8/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L144

**Comment by alamb** [comment]
> Ok, been a bit busy the past couple of days, continue to read the related codes and think a relatively good way to solve this today... One simple way I could think to solve it temporarily for not blocking the progress of #10943 is that we just treat it a special case like `count`:
> 
> https://github.com/apache/datafusion/blob/1e0c06e14ae821ac6aa344f8acb638431a898ae8/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L144

no worries -- there is a lot going on these days.

I think your short term workaround sounds very clever to me 👍  I think it is fine to leave figuring out the right general API to add to AggregateUdfImpl as a future project (my rationale being that special casing "max" / "min" is no worse than hard coding the `Min` and `Max` types, and unblocks the conversion to AggregateUDF)

**Comment by Rachelint** [comment]
> > Ok, been a bit busy the past couple of days, continue to read the related codes and think a relatively good way to solve this today... One simple way I could think to solve it temporarily for not blocking the progress of #10943 is that we just treat it a special case like `count`:
> > https://github.com/apache/datafusion/blob/1e0c06e14ae821ac6aa344f8acb638431a898ae8/datafusion/core/src/physical_optimizer/aggregate_statistics.rs#L144
> 
> no worries -- there is a lot going on these days.
> 
> I think your short term workaround sounds very clever to me 👍 I think it is fine to leave figuring out the right general API to add to AggregateUdfImpl as a future project (my rationale being that special casing "max" / "min" is no worse than hard coding the `Min` and `Max` types, and unblocks the conversion to AggregateUDF)

 :smile: Thanks, I just impl the short term workaround now.
But still worry about using the function name only to Identify the built-in `min/max/count` in the optimizer.
Assume that user define a to udf with a same name to overwrite the built-ins (seems we can do it?).
And the logic in user's `min/max/count` is totally different with the built-in. 
However the optimizer is unable to know the `min/max/count` is not the built-in(optimizer just identify by name), and do the mistaken optmization in result.

**Comment by alamb** [comment]
> 😄 Thanks, I just impl the short term workaround now. But still worry about using the function name only to identify the built-in `min/max/count` in the optimizer. Assume that user define a udaf with a same name to overwrite the built-in (seems we can do it?). And the logic in user's `min/max/count` is totally different with the built-in. However the optimizer is unable to know the `min/max/count` is not the built-in(optimizer just identify by name), and do the mistaken optmization in result.

I agree -- it will not be correct. I think that is why we eventually need to move the logic out of the optimizer and into the AggregateUDFImpl itself.

**Comment by Rachelint** [comment]
> > 😄 Thanks, I just impl the short term workaround now. But still worry about using the function name only to identify the built-in `min/max/count` in the optimizer. Assume that user define a udaf with a same name to overwrite the built-in (seems we can do it?). And the logic in user's `min/max/count` is totally different with the built-in. However the optimizer is unable to know the `min/max/count` is not the built-in(optimizer just identify by name), and do the mistaken optmization in result.
> 
> I agree -- it will not be correct. I think that is why we eventually need to move the logic out of the optimizer and into the AggregateUDFImpl itself.

🤔 Seems indeed necessary to do that... 
😄 Have finished the short term solution here.
I originally want to avoid `downcast_ref::<AggregateFunctionExpr>()`, but found if we do that, we should add two functions `is_distinct` and `function_name` to `AggregateExpr` before. I am not sure if it is ok to add them... so I just impl with `downcast_ref::<AggregateFunctionExpr>()` now...

File: datafusion/core/src/physical_optimizer/aggregate_statistics.rs [hunk 1]
```diff
@@ -273,6 +263,44 @@ fn take_optimizable_max(
     None
 }
 
+fn is_non_distinct_count(agg_expr: &dyn AggregateExpr) -> bool {
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/aggregate_statistics.rs:266]: I really like the encapsulation of this logic into functions 👍 

```suggestion
// TODO: Move this check into AggregateUDFImpl
// https://github.com/apache/datafusion/issues/11153
fn is_non_distinct_count(agg_expr: &dyn AggregateExpr) -> bool {
```

**Review by alamb — APPROVED** [review]
Thank you @Rachelint  -- I think this is an improvement as it will upblock pulling out Min/Max aggregate functions, even thought here is more to do. I also happen to think this PR is easier to understand the intent (by moving `is_min` and `is_max` checks into functions). 

Nice work

my only suggestion is to add comments with a link to the ticket that explains the rationale

**Comment by Rachelint** [comment]
Thanks @alamb for review, will continue to try to move them into `AggregateUDFImpl`, and comments have been added.

**Review by alamb — APPROVED** [review]
Thanks again @Rachelint  -- 🚀  we are making progress

**Comment by edmondop** [comment]
Thank you @Rachelint . I think I can resume https://github.com/apache/datafusion/pull/11013 unless @alamb you think there is more work we want to externalize/handle via separate PRs

**Comment by edmondop** [comment]
Right ! It is looking good. The refactoring has broken an optimiser, will need to investigate

**Comment by edmondop** [comment]
Makes sense. I am almost done on the other pr I need to understand why one optimisation was lost

**Pull request merged**